### DISCUSSION TITLE
Framework-side enums localisation

### DISCUSSION THREAD

**Discussion by Feodor0090** [ppy/osu#22226] [Q&A]
While most of settings are localised we still have these:
![изображение](https://user-images.githubusercontent.com/53872073/212572155-2769ea8b-5713-484e-8bae-140034ae8173.png)
These enums are coming from framework.

I can see 4 possible ways to deal with them.

- Open new "branch" on crowdin for framework alongside with web/client and pull strings for them from there. Not sure that it's possible because i didn't investigate does framework know about selected language in game or not.
- Remap them similar to `ManiaScrollingDirection` and localize these dublicates. Keep 2 bindables (from config and for dropdown) in section class for each and convert values between them there.
- Make custom dropdowns and override items creation there.
- Decide that threading mode, window mode and fps limiter options are non-translatable terms and leave them as is.

I would like to hear devs' opinion about this, i.e. what should be done.

**Comment by bdach** [comment]
There's another option here, and it's what I'd do: add extension methods for these enums, as in

```csharp
public static LocalisableString ToLocalisableString(this WindowMode windowMode)
{
    switch (windowMode)
    {
        case WindowMode.Windowed:
           return GraphicsSettingsStrings.ModeWindowed;

        // ...
    }
}
```

Doesn't involve framework at all. Probably would want @ppy/area-client to confirm first though.

**Reply by peppy** [reply]
Sounds best for now.

**Reply by smoogipoo** [reply]
In general I agree, but does this also mean that each `EnumDropdown` needs an override to call this particular method for the item text?

**Reply by bdach** [reply]
Good point. Maybe we can also provide an `EnumDropdown<T>(Func<T, LocalisableString>)` kind of ctor, and pass the extension method (or just static method at that point even) as a method group delegate to avoid subclassing too much.

As in:

```csharp
var dropdown = new OsuEnumDropdown<WindowMode>(FrameworkLocalisationExtensions.ToLocalisableString);
```

or something in that wheelhouse. There's probably some more work needed to fill out that picture.

**Reply by smoogipoo** [reply]
I wouldn't be against that, or as a property/ctor param on `Dropdown` itself.

**Comment by frenzibyte** [comment]
https://github.com/ppy/osu-framework/pull/6075, for reference.
